interfaces/builtin/custom_device.go

```diff
@@ -0,0 +1,367 @@
+		}
+	}
+
+	if err := iface.validatePaths(slot, "read"); err != nil {
+		return err
+	}
+	if err := iface.validatePaths(slot, "write"); err != nil {
+		return err
+	}
+
+	var udevTaggingRules []map[string]interface{}
+	err = slot.Attr("udev-tagging", &udevTaggingRules)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return err
+	}
+	for _, udevTaggingRule := range udevTaggingRules {
+		if err := iface.validateUDevTaggingRule(udevTaggingRule, devices); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	customDeviceAttr, isSet := plug.Attrs["custom-device"]
+	customDevice, ok := customDeviceAttr.(string)
+	if isSet && !ok {
+		return fmt.Errorf(`custom-device "custom-device" attribute must be a string, not %v`,
+			plug.Attrs["custom-device"])
+	}
+	if customDevice == "" {
+		if plug.Attrs == nil {
+			plug.Attrs = make(map[string]interface{})
+		}
+		// custom-device defaults to "plug" name if unspecified
+		plug.Attrs["custom-device"] = plug.Name
+	}
+
+	return nil
+}
+
+func (iface *customDeviceInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	snippet := &bytes.Buffer{}
+	emitRule := func(paths []string, permissions string) {
+		for _, path := range paths {
+			fmt.Fprintf(snippet, "\"%s\" %s,\n", path, permissions)
+		}
+	}
+
+	// get all attributes without validation, since that was done before;
+	// should an error occur, we'll simply not write any rule.
+
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+	emitRule(devicePaths, "rw")
+
+	var writablePaths []string
+	_ = slot.Attr("write", &writablePaths)
+	emitRule(writablePaths, "rw")
+
+	var readOnlyPaths []string
+	_ = slot.Attr("read", &readOnlyPaths)
+	emitRule(readOnlyPaths, "r")
+
+	spec.AddSnippet(snippet.String())
+	return nil
+}
+
+// extractStringMapAttribute looks up the given key in the container, and
+// returns its value as a map[string]string.
+// No validation is performed, since it already occurred before connecting the
+// interface.
+func (iface *customDeviceInterface) extractStringMapAttribute(container map[string]interface{}, key string) map[string]string {
+	valueMap, ok := container[key].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	stringMap := make(map[string]string, len(valueMap))
+	for key, value := range valueMap {
+		stringMap[key] = value.(string)
+	}
+
+	return stringMap
+}
+
+func (iface *customDeviceInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var devicePaths []string
+	_ = slot.Attr("devices", &devicePaths)
+
+	deviceRules := make(map[string]string, len(devicePaths))
+	for _, devicePath := range devicePaths {
+		if strings.HasPrefix(devicePath, "/dev/") {
+			deviceName := devicePath[5:]
+			deviceRule := fmt.Sprintf(`KERNEL=="%s"`, deviceName)
```

Comment:
do we have a test that checks this default rule? I looked at the tests and I'm not sure from a quick scan

Reply:
Yes, the `TestUDevSpec` test verifies that a udev rule is generated for the `/dev/input/event[0-9]` device, which does not have a `udev-tagging` stanza associated to it.